lib/ansible/playbook/task.py

```diff
@@ -118,18 +118,34 @@ def get_path(self):
     def get_name(self, include_role_fqcn=True):
         ''' return the name of the task '''
 
-        if self._role:
-            role_name = self._role.get_name(include_role_fqcn=include_role_fqcn)
+        role_name = None
+        role_fqcn_name = None
 
-        if self._role and self.name and role_name not in self.name:
-            return "%s : %s" % (role_name, self.name)
-        elif self.name:
-            return self.name
-        else:
-            if self._role:
-                return "%s : %s" % (role_name, self.action)
+        if not self._role:
+            if self.name:
+                return "%s" % self.name
             else:
-                return "%s" % (self.action,)
+                return "%s" % self.action
+
+        role_name = self._role.get_name(include_role_fqcn=False)
+
+        if include_role_fqcn:
+            role_fqcn_name = self._role.get_name(include_role_fqcn=True)
+
+        if self.name:
+            if role_fqcn_name and not self.name.startswith("%s : " % role_fqcn_name):
+                return "%s : %s" % (role_fqcn_name, self.name)
+            if role_name and not self.name.startswith("%s : " % role_name):
```

Comment:
Since self.name is user-defined, is there a benefit to checking for strings in it rather than just always prepending the role name? What is the expected behavior if a handler's name defined at the start of the main play starts with `role_name : ` and a handler name in a role also starts with `role_name : `? I think I'd still expect for the handler in the role to be prefix to be able to distinguish it from the other handler.

Related to that, I'm not sure if it's a bug (though it seems like the intention was to give the user a way to prefer the role's namespace), but this code uses definition order rather than verifying there is not a matching fully qualified handler name before falling back to a less exact candidate. https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/strategy/__init__.py#L486 So if someone has a handler defined at the start of the play named 'role_name : something' and a handler named 'something' in the role role_name, what should `notify: 'role_name : something'` do inside the role? I don't know, but I think we might want to scope this out and document it to help prevent breakage in the future.

Reply:
Hm... I think the original reason for the string checking logic was for the callback.

Reply:
> Since self.name is user-defined, is there a benefit to checking for strings in it rather than just always prepending the role name? What is the expected behavior if a handler's name defined at the start of the main play starts with `role_name : ` and a handler name in a role also starts with `role_name : `? I think I'd still expect for the handler in the role to be prefix to be able to distinguish it from the other handler.

Not sure if I understand correctly, but if a handler has been defined with a duplicate name, the first one defined will be notified.
This is afaik one of the reasons to prefix the handler name with `role_name : `.
So if you want to distinguish the handlers, use `role_name : ` for the one defined in the play and `role_name : role_name : ` to call the role defined handler. (If the handler name starts with `role_name : ` it becomes a bit awkward, but that seems like designed)

Reply:
Yeah, that is the current behavior. The precedence of notification is sort of tangential to my thoughts about your change (not suggesting you change that code), but the behavior on devel is different than the original implementation (I'm using 8218591fecf124f3bb8db22030959e235b450f02 as the source of truth for how it was 'supposed' to work), so it just adds another layer to consider. I think this method is overloaded. Before that commit, it looks like get_name was really just used for debugging output and the callback, and then that commit used it for the handler search.

> (If the handler name starts with role_name : it becomes a bit awkward, but that seems like designed)

I think it's ambiguous if it's like designed. My rambling might be entirely wrong, but the more I look at it, the more I don't think https://github.com/ansible/ansible/commit/8218591fecf124f3bb8db22030959e235b450f02#diff-ac960c5e7aa7ed715ac443f64451bff6R108 was supposed to have anything to do with the handler search really; I think it was added just to avoid displaying an additional `role_name : ` for the callback. The string checking seems fragile so maybe callbacks should use a distinct path, and then everything else can just use the one-liner fix `if self._role and self.name:`.

Reply:
Alright, I think I get what you're trying to say.

This PR was an attempt to try to limit the changes to the then-current design, while allowing the role name to be part of the handler name, as I don't feel comfortable enough with the whole process/codebase to propose other changes without some guidance.  
So feel free to contact me on IRC/matrix if you want to discuss this more. I'll join #ansible-devel right away.